### PR TITLE
[Lint][AMDGPU] No store to const addrspace

### PR DESCRIPTION
**Pull request by jofrn**
Ensure store to const addrspace is not allowed by Linter.

### CODE REVIEW

File: llvm/lib/IR/Verifier.cpp [hunk 1]
```diff
@@ -4224,6 +4224,33 @@ void Verifier::visitLoadInst(LoadInst &LI) {
   visitInstruction(LI);
 }
 
+static bool isConstantAddressSpace(unsigned AS) {
```
Comment (shiltian) [inline llvm/lib/IR/Verifier.cpp:4227]: constant address space is not an AMD only thing. I'm thinking probably we can split this function into target dependent (for example, to query whether an AS is a constant AS) and independent part.

Comment (jofrn) [reply llvm/lib/IR/Verifier.cpp:4227]: What would the independent part contain if we do not have a use case for it yet?

**Review by arsenm — CHANGES_REQUESTED** [review]
This must be UB, and not a verifier error. For example, we allow constant->flat casts. We do not want to have to consider the address spaces involved when folding something like store (addrspacecast ptr addrspace(4) %val to ptr)

**Comment by arsenm** [comment]
Plus long term I don't think we should have a constant address space at all.

**Comment by jofrn** [comment]
> Was there any discussion about this? I'm not sure if it should be a verifier error or just undefined behaviour at run time. Also there are other instructions that write to memory, like atomicrmw.
It can be guarded with a flag perhaps. There was discussion to catch this in some way or another since it leads to more confusing error messages later on.

**Comment by nikic** [comment]
These kinds of checks should go into the Lint pass instead of the Verifier pass.

**Comment by arsenm** [comment]
> The addrspace in `store (addrspacecast ptr addrspace(4) %val to ptr)` won't be affected by `getPointerAddressSpace` since this function asks about the second operand to the store, `store i32 %val, ptr addrspace(4) %ptr`, in particular, `ptr addrspace(4) %ptr`. Your example gives a complex expression on the value source, the first operand.

You're reading the example too literally. It should be valid for a pass to strip away the cast, and store directly to the original address space of the pointer without considering any special information. It would be a target specific burden placed on any IR transformation, which is not OK.

This type of check belongs in Lint, which hardly ever gets used and I'm not sure is worth the effort

File: llvm/test/CodeGen/AMDGPU/const-store.ll [hunk 2]
```diff
@@ -0,0 +1,9 @@
+; RUN: not opt -mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o - |& FileCheck %s
+
+define amdgpu_kernel void @store_const(ptr addrspace(4) %out, i32 %a, i32 %b) addrspace(4) {
+; CHECK: Undefined behavior: Write to const memory
+; CHECK-NEXT: store i32 %r, ptr addrspace(4) %out
+  %r = add i32 %a, %b
+  store i32 %r, ptr addrspace(4) %out
+  ret void
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/const-store.ll:9]: test atomicrmw, cmpxchg, and a memset, memcpy? Also at least the addrspace(6) case

File: llvm/test/CodeGen/AMDGPU/const-store.ll [hunk 3]
```diff
@@ -0,0 +1,9 @@
+; RUN: not opt -mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o - |& FileCheck %s
+
+define amdgpu_kernel void @store_const(ptr addrspace(4) %out, i32 %a, i32 %b) addrspace(4) {
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/const-store.ll:3]: no reason to set the address space of the function

File: llvm/test/CodeGen/AMDGPU/const-store.ll [hunk 4]
```diff
@@ -0,0 +1,9 @@
+; RUN: not opt -mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o - |& FileCheck %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/const-store.ll:2]: negative test with a different target

File: llvm/test/CodeGen/AMDGPU/const-store.ll [hunk 5]
```diff
@@ -0,0 +1,9 @@
+; RUN: not opt -mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o - |& FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/const-store.ll:1]: Should move this to test/Analysis/Lint

File: llvm/lib/Analysis/Lint.cpp [hunk 6]
```diff
@@ -378,6 +382,36 @@ void Lint::visitReturnInst(ReturnInst &I) {
   }
 }
 
+bool Lint::isConstantAddressSpace(unsigned AS) const {
+  if (TT.isAMDGPU()) {
+    switch(AS) {
+      using namespace AMDGPUAS;
+    case CONSTANT_ADDRESS:
```
Comment (nikic) [inline llvm/lib/Analysis/Lint.cpp:389]: Should this be a helper in AMDGPUAddrSpace.h? I'd expect this logic to already exist somewhere, but it looks like AMDGPUAA is only looking for CONSTANT_ADDRESS and nothing else: https://github.com/llvm/llvm-project/blob/97aa8cc94d94e8f0adc85489f7832ba7c0a9b577/llvm/lib/Target/AMDGPU/AMDGPUAliasAnalysis.cpp#L78

**Review by nikic — COMMENTED** [review]
New approach looks reasonable to me.

File: llvm/lib/Analysis/Lint.cpp [hunk 7]
```diff
@@ -401,6 +403,10 @@ void Lint::visitMemoryReference(Instruction &I, const MemoryLocation &Loc,
         "Unusual: Address one pointer dereference", &I);
 
   if (Flags & MemRef::Write) {
+    if (TT.isAMDGPU())
+      Check(!AMDGPU::isConstantAddressSpace(UnderlyingObject->getType()->getPointerAddressSpace()),
```
Comment (shiltian) [inline llvm/lib/Analysis/Lint.cpp:407]: Is it possible or better to do the check via TTI? I'm not sure if it's a good idea to do the check based on triple.

File: llvm/test/Analysis/Lint/const-store.ll [hunk 8]
```diff
@@ -0,0 +1,21 @@
+; RUN: not opt --mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o /dev/null |& FileCheck %s
+; RUN: opt --mtriple=amdgcn --mcpu=gfx1030 --passes=lint %s -o /dev/null |& FileCheck %s --check-prefixes=CHECK,CHECK0
+; RUN: opt --mtriple=x86_64 --passes=lint --lint-abort-on-error %s -o /dev/null |& FileCheck %s --allow-empty --check-prefix=NOERR
+; NOERR: {{^$}}
+
```
Comment (shiltian) [inline llvm/test/Analysis/Lint/const-store.ll:5]: I think you will need `; REQUIRES: amdgpu-registered-target`.

File: llvm/lib/Analysis/Lint.cpp (hunk 7 above)
Comment (shiltian) [reply llvm/lib/Analysis/Lint.cpp:409]: Just to be clear, the "check" I meant here was `AMDGPU::isConstantAddressSpace`.

File: llvm/include/llvm/Support/AMDGPUAddrSpace.h [hunk 9]
```diff
@@ -93,6 +93,33 @@ inline bool isExtendedGlobalAddrSpace(unsigned AS) {
          AS == AMDGPUAS::CONSTANT_ADDRESS_32BIT ||
          AS > AMDGPUAS::MAX_AMDGPU_ADDRESS;
 }
+
+inline bool isConstantAddressSpace(unsigned AS) {
+  switch (AS) {
+    using namespace AMDGPUAS;
```
Comment (arsenm) [inline llvm/include/llvm/Support/AMDGPUAddrSpace.h:99]:
```suggestion
  using namespace AMDGPUAS;
  switch (AS) {
```

File: llvm/test/Analysis/Lint/const-store.ll (hunk 8 above)
Comment (arsenm) [reply llvm/test/Analysis/Lint/const-store.ll:7]: This test doesn't actually require the targets. The triple will still be recognized no matter what, which is all that's needed

File: llvm/test/Analysis/Lint/const-store.ll [hunk 10]
```diff
@@ -0,0 +1,42 @@
+; RUN: not opt --mtriple=amdgcn --passes=lint --lint-abort-on-error %s -o /dev/null |& FileCheck %s
+; RUN: opt --mtriple=amdgcn --mcpu=gfx1030 --passes=lint %s -o /dev/null |& FileCheck %s --check-prefixes=CHECK,CHECK0
+; RUN: opt --mtriple=x86_64 --passes=lint --lint-abort-on-error %s -o /dev/null |& FileCheck %s --allow-empty --check-prefix=NOERR
```
Comment (arsenm) [inline llvm/test/Analysis/Lint/const-store.ll:3]: Should use -disable-output instead of -o /dev/null. Does |& work on all shells? Usually use 2>&1

**Review by arsenm — APPROVED** [review]
lgtm with nit

**Pull request merged**